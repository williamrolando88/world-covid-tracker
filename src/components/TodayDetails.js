import { ArrowCircleLeftIcon } from '@heroicons/react/outline';
import { MenuIcon, MicrophoneIcon } from '@heroicons/react/solid';
import React from 'react';
import { useSelector } from 'react-redux';
import { Navigate, useNavigate } from 'react-router-dom';
import ErrorScreen from './ErrorScreen';
import LoadingScreen from './LoadingScreen';

const TodayDetails = () => {
  const navigate = useNavigate();
  const { status, data } = useSelector(
    (store) => store.covidData.countrySingleDay,
  );

  if (!status) {
    return <Navigate to="/" />;
  }

  const formatNumber = (number) => new Intl.NumberFormat().format(number);

  const totalStats = {
    Confirmed: { value: 0 },
    Recovered: { value: 0 },
    'Open Cases': { value: 0 },
    Deaths: { value: 0 },
  };

  if (status === 'fulfilled') {
    totalStats.Confirmed.value = data.today_confirmed;
    totalStats.Recovered.value = data.today_recovered;
    totalStats['Open Cases'].value = data.today_open_cases;
    totalStats.Deaths.value = data.today_deaths;
  }

  return status === 'pending' ? (
    <LoadingScreen />
  ) : status === 'fulfilled' ? (
    <div className="h-screen bg-[#24243e]">
      {/* Navbar */}
      <nav className="font-gill-sans flex items-center justify-between bg-gradient-to-b from-[#0f0c29] via-[#302b63]  to-[#24243e] px-2 py-2 text-white">
        <button
          type="button"
          className="flex items-center rounded-md border py-1 px-2 uppercase"
          onClick={() => navigate(-1)}
        >
          <ArrowCircleLeftIcon className="h-6 w-6" />
          <span>BACK</span>
        </button>
        <h1 className="text-lg capitalize">Country details</h1>
        <div className="flex items-center gap-2">
          <button className="h-6 w-6" type="button">
            <MicrophoneIcon />
          </button>
          <button className="h-6 w-6" type="button">
            <MenuIcon />
          </button>
        </div>
      </nav>
      {/* Main container */}
      <article className="font-lato flex flex-col items-stretch text-center text-slate-100">
        {/* Header */}
        <section className="h-48 w-screen bg-[url(./assets/covid.jpg)] bg-cover bg-center bg-no-repeat text-white">
          <div className="flex h-full flex-col items-center bg-black/60">
            <div className="grid grow content-center">
              <h2 className="font-gill-sans text-center text-2xl">
                {data.name}
              </h2>
            </div>
            <p className="m-1 self-end text-xs">{`Source: ${data.source}`}</p>
          </div>
        </section>
        {/* Separator */}
        <p className="font-gill-sans bg-gradient-to-b from-[#0f0c29] via-[#302b63] to-[#24243e] py-1 text-center text-sm  uppercase">{`Today's data (${data.date})`}</p>
        {/* Total data */}
        <div className="flex justify-evenly bg-[#302b63] py-2">
          {Object.keys(totalStats).map((key) => (
            <div key={key} className="flex flex-col">
              <h4>{formatNumber(totalStats[key].value)}</h4>
              <h3 className="font-gill-sans text-xs capitalize">
                {`Total
                ${key}`}
              </h3>
            </div>
          ))}
        </div>
        {/* Growing rate */}
        <div className="grid grid-cols-2">
          <div className="bg-[url(./assets/total.webp)] bg-contain bg-center bg-no-repeat">
            <div className="flex h-36 flex-col items-stretch justify-center bg-[#302b63]/80">
              <h3>Total New Confirmed</h3>
              <h4>{formatNumber(data.today_new_confirmed)}</h4>
            </div>
          </div>
          <div className="bg-[url(./assets/recovered.webp)] bg-contain bg-center bg-no-repeat">
            <div className="flex h-36 flex-col items-stretch justify-center bg-[#24243e]/80">
              <h3>Total New Recovered</h3>
              <h4>{formatNumber(data.today_new_recovered)}</h4>
            </div>
          </div>
          <div className="bg-[url(./assets/confirmed.webp)] bg-contain bg-center bg-no-repeat">
            <div className="flex h-36 flex-col items-stretch justify-center bg-[#24243e]/80">
              <h3>Total New Open Cases</h3>
              <h4>{formatNumber(data.today_new_open_cases)}</h4>
            </div>
          </div>
          <div className="bg-[url(./assets/dead.webp)] bg-contain bg-center bg-no-repeat">
            <div className="flex h-36 flex-col items-stretch justify-center bg-[#302b63]/80">
              <h3>Total New Deaths</h3>
              <h4>{formatNumber(data.today_new_deaths)}</h4>
            </div>
          </div>
        </div>
      </article>
    </div>
  ) : (
    <ErrorScreen />
  );
};

export default TodayDetails;
