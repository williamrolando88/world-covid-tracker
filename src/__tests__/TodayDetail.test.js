import configureStore from 'redux-mock-store';
import { Provider } from 'react-redux';
import { BrowserRouter } from 'react-router-dom';
import renderer from 'react-test-renderer';
import TodayDetails from '../components/TodayDetails';
import { render } from '@testing-library/react';

const initialState = {
  covidData: {
    countrySingleDay: {
      status: 'fulfilled',
      data: {
        today_confirmed: 1000,
        today_recovered: 2000,
        today_open_cases: 3000,
        today_deaths: 4000,
      },
    },
  },
};

const mockStore = configureStore();

describe('Should render', () => {
  test('should render TodayDetails', () => {
    const details = renderer.create(
      <Provider store={mockStore(initialState)}>
        <BrowserRouter>
          <TodayDetails />
        </BrowserRouter>
      </Provider>,
    );
    expect(details).toMatchSnapshot();
  });
});

describe('should render the given values', () => {
  test('should render total_confirmed', () => {
    const { getByText } = render(
      <Provider store={mockStore(initialState)}>
        <BrowserRouter>
          <TodayDetails />
        </BrowserRouter>
      </Provider>,
    );
    getByText('1.000');
    getByText('2.000');
    getByText('3.000');
    getByText('4.000');
  });
});
