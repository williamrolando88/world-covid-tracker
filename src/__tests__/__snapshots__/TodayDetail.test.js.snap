// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render should render TodayDetails 1`] = `
<div
  className="h-screen bg-[#24243e]"
>
  <nav
    className="font-gill-sans flex items-center justify-between bg-gradient-to-b from-[#0f0c29] via-[#302b63]  to-[#24243e] px-2 py-2 text-white"
  >
    <button
      className="flex items-center rounded-md border py-1 px-2 uppercase"
      onClick={[Function]}
      type="button"
    >
      <svg
        aria-hidden="true"
        className="h-6 w-6"
        fill="none"
        stroke="currentColor"
        viewBox="0 0 24 24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M11 15l-3-3m0 0l3-3m-3 3h8M3 12a9 9 0 1118 0 9 9 0 01-18 0z"
          strokeLinecap="round"
          strokeLinejoin="round"
          strokeWidth={2}
        />
      </svg>
      <span>
        BACK
      </span>
    </button>
    <h1
      className="text-lg capitalize"
    >
      Country details
    </h1>
    <div
      className="flex items-center gap-2"
    >
      <button
        className="h-6 w-6"
        type="button"
      >
        <svg
          aria-hidden="true"
          fill="currentColor"
          viewBox="0 0 20 20"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            clipRule="evenodd"
            d="M7 4a3 3 0 016 0v4a3 3 0 11-6 0V4zm4 10.93A7.001 7.001 0 0017 8a1 1 0 10-2 0A5 5 0 015 8a1 1 0 00-2 0 7.001 7.001 0 006 6.93V17H6a1 1 0 100 2h8a1 1 0 100-2h-3v-2.07z"
            fillRule="evenodd"
          />
        </svg>
      </button>
      <button
        className="h-6 w-6"
        type="button"
      >
        <svg
          aria-hidden="true"
          fill="currentColor"
          viewBox="0 0 20 20"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            clipRule="evenodd"
            d="M3 5a1 1 0 011-1h12a1 1 0 110 2H4a1 1 0 01-1-1zM3 10a1 1 0 011-1h12a1 1 0 110 2H4a1 1 0 01-1-1zM3 15a1 1 0 011-1h12a1 1 0 110 2H4a1 1 0 01-1-1z"
            fillRule="evenodd"
          />
        </svg>
      </button>
    </div>
  </nav>
  <article
    className="font-lato flex flex-col items-stretch text-center text-slate-100"
  >
    <section
      className="h-48 w-screen bg-[url(./assets/covid.jpg)] bg-cover bg-center bg-no-repeat text-white"
    >
      <div
        className="flex h-full flex-col items-center bg-black/60"
      >
        <div
          className="grid grow content-center"
        >
          <h2
            className="font-gill-sans text-center text-2xl"
          />
        </div>
        <p
          className="m-1 self-end text-xs"
        >
          Source: undefined
        </p>
      </div>
    </section>
    <p
      className="font-gill-sans bg-gradient-to-b from-[#0f0c29] via-[#302b63] to-[#24243e] py-1 text-center text-sm  uppercase"
    >
      Today's data (undefined)
    </p>
    <div
      className="flex justify-evenly bg-[#302b63] py-2"
    >
      <div
        className="flex flex-col"
      >
        <h4>
          1.000
        </h4>
        <h3
          className="font-gill-sans text-xs capitalize"
        >
          Total
                Confirmed
        </h3>
      </div>
      <div
        className="flex flex-col"
      >
        <h4>
          2.000
        </h4>
        <h3
          className="font-gill-sans text-xs capitalize"
        >
          Total
                Recovered
        </h3>
      </div>
      <div
        className="flex flex-col"
      >
        <h4>
          3.000
        </h4>
        <h3
          className="font-gill-sans text-xs capitalize"
        >
          Total
                Open Cases
        </h3>
      </div>
      <div
        className="flex flex-col"
      >
        <h4>
          4.000
        </h4>
        <h3
          className="font-gill-sans text-xs capitalize"
        >
          Total
                Deaths
        </h3>
      </div>
    </div>
    <div
      className="grid grid-cols-2"
    >
      <div
        className="bg-[url(./assets/total.webp)] bg-contain bg-center bg-no-repeat"
      >
        <div
          className="flex h-36 flex-col items-stretch justify-center bg-[#302b63]/80"
        >
          <h3>
            Total New Confirmed
          </h3>
          <h4>
            NaN
          </h4>
        </div>
      </div>
      <div
        className="bg-[url(./assets/recovered.webp)] bg-contain bg-center bg-no-repeat"
      >
        <div
          className="flex h-36 flex-col items-stretch justify-center bg-[#24243e]/80"
        >
          <h3>
            Total New Recovered
          </h3>
          <h4>
            NaN
          </h4>
        </div>
      </div>
      <div
        className="bg-[url(./assets/confirmed.webp)] bg-contain bg-center bg-no-repeat"
      >
        <div
          className="flex h-36 flex-col items-stretch justify-center bg-[#24243e]/80"
        >
          <h3>
            Total New Open Cases
          </h3>
          <h4>
            NaN
          </h4>
        </div>
      </div>
      <div
        className="bg-[url(./assets/dead.webp)] bg-contain bg-center bg-no-repeat"
      >
        <div
          className="flex h-36 flex-col items-stretch justify-center bg-[#302b63]/80"
        >
          <h3>
            Total New Deaths
          </h3>
          <h4>
            NaN
          </h4>
        </div>
      </div>
    </div>
  </article>
</div>
`;
